dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java

```diff
@@ -0,0 +1,144 @@
+  private final DoraMetaStore mMetastore;
+  private final CacheManager mCacheManager;
+  private final PagedDoraWorker mDoraWorker;
+  private final UnderFileSystem mUfs;
+  private final LockPool<String> mPathLocks;
+  private final long mPageSize =
+      Configuration.global().getBytes(PropertyKey.WORKER_PAGE_STORE_PAGE_SIZE);
+
+  public DoraMetaManager(
+      PagedDoraWorker doraWorker, DoraMetaStore metaStore, CacheManager cacheManger,
+      UnderFileSystem ufs, LockPool<String> pathLocks) {
+    mMetastore = metaStore;
+    mCacheManager = cacheManger;
+    mDoraWorker = doraWorker;
+    mUfs = ufs;
+    mPathLocks = pathLocks;
+  }
+
+  /**
+   * Gets file meta from UFS.
+   * @param path the full ufs path
+   * @return the file status, or empty optional if not found
+   */
+  public Optional<FileStatus> getFromUfs(String path) throws IOException {
+    try {
+      UfsStatus status = mUfs.getStatus(path);
+      DoraMeta.FileStatus fs = mDoraWorker.buildFileStatusFromUfsStatus(status, path);
+      return Optional.ofNullable(fs);
+    } catch (FileNotFoundException e) {
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Gets file meta from UFS and loads it into metastore if exists.
+   * @param path the full ufs path
+   * @return the file status, or empty optional if not found
+   */
+  public Optional<FileStatus> loadFromUfs(String path) throws IOException {
+    Optional<FileStatus> fileStatus = getFromUfs(path);
+    try (var lockResource = mPathLocks.get(path, LockMode.WRITE)) {
+      fileStatus.ifPresent(status -> put(path, status));
+      return fileStatus;
+    }
+  }
+
+  /**
+   * Gets file meta from the metastore.
+   * @param path the full ufs path
+   * @return the file status, or empty optional if not found
+   */
+  public Optional<FileStatus> getFromMetaStore(String path) {
+    try (var lockResource = mPathLocks.get(path, LockMode.READ)) {
+      return mMetastore.getDoraMeta(path);
+    }
+  }
+
+  /**
+   * Puts meta of a file into the metastore, and invalidates the file data cache.
+   * @param path the full ufs path
+   * @param meta the file meta
+   */
+  public void put(String path, FileStatus meta) {
+    try (var lockResource = mPathLocks.get(path, LockMode.WRITE)) {
+      Optional<FileStatus> status = mMetastore.getDoraMeta(path);
+      if (status.isPresent() && !status.get().getFileInfo().getFolder()
+          && status.get().getFileInfo().getLength() > 0) {
+        Fingerprint originalFingerprint =
+            Fingerprint.parse(status.get().getFileInfo().getUfsFingerprint());
+        if (originalFingerprint == null) {
+          originalFingerprint = Fingerprint.INVALID_FINGERPRINT;
+        }
+        Fingerprint newFingerprint =
+            Fingerprint.parse(meta.getFileInfo().getUfsFingerprint());
+        if (newFingerprint == null) {
+          newFingerprint = Fingerprint.INVALID_FINGERPRINT;
+        }
+        boolean contentMatched = originalFingerprint.isValid()
```

Comment:
this is an optimization that compares the fingerprint and if the file content (mainly based of off the content hash) does not change, we do not invalidate the file cache

Reply:
will add some tests for this one